manticore/platforms/linux.py

```diff
@@ -692,14 +692,23 @@ def __init__(
         # Keep track of the symbolic inputs we create
         self.inputs_recvd: List[ArrayProxy] = []
         self.recv_pos = 0
+        # This is a meta-variable, of sorts, and it is responsible for
+        # determining the symbolic length of the array during recv/read.
+        # Initially, it is None, to indicate we haven't forked yet. After
+        # fork, each state will be assigned their respective _actual_,
+        # concretized, receive length
+        self._symb_len: Optional[int] = None
+        # Set after adding this socket to the file descriptor list
+        self.fd: Optional[int] = None
 
     def __getstate__(self):
         state = super().__getstate__()
         state["inputs_recvd"] = self.inputs_recvd
         state["symb_name"] = self.symb_name
         state["recv_pos"] = self.recv_pos
         state["max_recv_symbolic"] = self.max_recv_symbolic
-        state["constraints"] = self._constraints
```

Comment:
Why'd you remove the constraint set from the pickled state? It seems to me like you'd want it to persist.

Reply:
Good question. I believe this should be taken care of during the `SLinux` `__setstate__` here https://github.com/trailofbits/manticore/blob/898918554998c47d624d8df28ff8d806bd9f49ce/manticore/platforms/linux.py#L3278-L3282

We also initialize a SymbolicSocket with a reference to the `ConstraintSet` in `SLinux` here https://github.com/trailofbits/manticore/blob/898918554998c47d624d8df28ff8d806bd9f49ce/manticore/platforms/linux.py#L3413

I'm open to suggestions for redesigning how/where the constraints are added. However, without removing this line, I think we are duplicating serialization of the constraint set, right?

Reply:
The problem with keeping links to the constraintset is forking.

This all needs a refactor/simplification. But currently we fork using a context manager: 
https://github.com/trailofbits/manticore/blob/898918554998c47d624d8df28ff8d806bd9f49ce/manticore/core/manticore.py#L504-L515

Note that there the State is spawned into a child state using the width statement. Supposedly modifying anything of the child state would not affect the parent. (This is a lie except for constraints)

Note that the child constraints reference/references are updated when spawning a child from a parent:
https://github.com/trailofbits/manticore/blob/898918554998c47d624d8df28ff8d806bd9f49ce/manticore/core/state.py#L150

If you keep a new reference to a constraintset it should be updated there.

Reply:
@feliam Can you check out this commit https://github.com/trailofbits/manticore/pull/1786/commits/aeb4a3100be2c1e7e7aafb3b8f516abd559cb371 if it would work as a fix for this? Thank you!

Reply:
Yes it looks like the new reference to the constraints are correctly propagated with that.

Reply:
@ehennenfent Anything else you're wondering about this after the latest commits?

Reply:
Nope, looks good to me! All this constraint set stuff is tricky...